### PR TITLE
Improved Non Sequential Models Support and New Layers 

### PR DESCRIPTION
**Pull request by pranv**
<b>Non Sequential Models:<b>

Each layer has 2 attributes: `name` and `prev`. 

* `name` is a string that internally maps to a layer number.
* `prev` is the string that internally maps to the layer that current layer should be connected to, also internally mapped to a layer number. Defaults the last added layer.

The `Layer` class takes inputs from multiple previous layers as a list. For backward compatibility, single input is handled.

```
model = Sequential()
model.add(Dense(20,5, name='input'))
model.add(Dense(5, 1,name='pro1', prev='input'))
model.add(Dense(5, 1, name='pro2', prev='input'))
model.add(Merge(prev=['pro1', 'pro2'], mode='concat'))
```

<b>Other Changes:</b>
* Added LRN Layer
* Added Pooling with Strides as supported by Theano 0.7
* Added ZeroPadding2D Layer

### CODE REVIEW

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -97,65 +99,82 @@ def __init__(self, nb_filter, stack_size, nb_row, nb_col,
...
-            "subsample":self.subsample}
+                "nb_filter":self.nb_filter,
+                "stack_size":self.stack_size,
+                "nb_row":self.nb_row,
+                "nb_col":self.nb_col,
+                "init":self.init.__name__,
+                "activation":self.activation.__name__,
+                "image_shape":self.image_shape,
+                "border_mode":self.border_mode,
+                "subsample":self.subsample}
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Why the added indents?

File: keras/layers/advanced_activations.py [hunk 2]
```diff
@@ -35,4 +37,4 @@ def get_output(self, train):
 
     def get_config(self):
         return {"name":self.__class__.__name__,
-        "input_shape":self.input_shape}
+        "input_dim":self.input_dim}
```
Comment (fchollet) [inline keras/layers/advanced_activations.py]: This attribute doesn't exist.

File: keras/layers/core.py [hunk 3]
```diff
@@ -16,21 +16,36 @@
 srng = RandomStreams(seed=np.random.randint(10e6))
 
 class Layer(object):
-    def __init__(self):
+    def __init__(self, name, prev):
```
Comment (fchollet) [inline keras/layers/core.py]: What is the purpose of passing the name of the layer and the name of the previous layer? Isn't the name of the previous layer readily accessible as `self.previous.name`?

File: keras/layers/core.py [hunk 4]
```diff
@@ -247,23 +211,24 @@ class Dense(Layer):
     '''
         Just your regular fully connected NN layer.
     '''
-    def __init__(self, input_dim, output_dim, init='glorot_uniform', activation='linear', weights=None, 
-        W_regularizer=None, b_regularizer=None, W_constraint=None, b_constraint=None):
+    def __init__(self, input_shape, output_shape,init='glorot_uniform', activation='linear', weights=None,
```
Comment (fchollet) [inline keras/layers/core.py]: In terms of code style, please use spaces after commas.

File: keras/layers/core.py [hunk 5]
```diff
@@ -323,115 +289,12 @@ def act_func(X):
...
     def get_config(self):
         return {"name":self.__class__.__name__,
-            "input_dim":self.input_dim,
-            "output_dim":self.output_dim,
+            "input_shape":self.input_shape,
+            "output_shape":self.output_shape,
             "init":self.init.__name__,
             "activation":self.activation.__name__}
 
-class AutoEncoder(Layer):
```
Comment (fchollet) [inline keras/layers/core.py]: Whis is the autoencoder going away?

File: keras/layers/normalization.py [hunk 6]
```diff
@@ -64,6 +65,42 @@ def get_output(self, train):
...
         return {"name":self.__class__.__name__,
-            "input_shape":self.input_shape,
+            "input_dim":self.input_dim,
             "epsilon":self.epsilon,
             "mode":self.mode}
+
+class LRN2D(Layer):
+    """
+    This code is adapted from pylearn2.
+    License at: https://
```
Comment (fchollet) [inline keras/layers/normalization.py]: License at?

File: keras/models.py [hunk 7]
```diff
@@ -100,28 +112,74 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
...
             allow_input_downcast=True, mode=theano_mode)
-        self._test = theano.function(test_ins, test_score, 
+        self._test = theano.function(test_ins, test_score,
             allow_input_downcast=True, mode=theano_mode)
-        self._test_with_acc = theano.function(test_ins, [test_score, test_accuracy], 
+        self._test_with_acc = theano.function(test_ins, [test_score, test_accuracy],
             allow_input_downcast=True, mode=theano_mode)
 
 
+class Sequential(Model):
```
Comment (fchollet) [inline keras/models.py]: In modern Keras, the Sequential model is build through double inheritance from the Sequential container and the generic Model. Why not here?

File: keras/models.py [hunk 8]
```diff
@@ -240,14 +297,14 @@ def predict(self, X, batch_size=128, verbose=1):
                 preds = np.zeros(shape)
             preds[batch_start:batch_end] = batch_preds
 
-            if verbose == 1:
+            if verbose==1:
```
Comment (fchollet) [inline keras/models.py]: PEP8: spaces around operators. We don't do strict PEP8, but such guidelines make the code more readable and should be respected.

File: keras/models.py [hunk 9]
```diff
@@ -296,32 +353,6 @@ def evaluate(self, X, y, batch_size=128, show_accuracy=False, verbose=1):
         else:
             return tot_score / seen
 
-
-class Sequential(Model, containers.Sequential):
```
Comment (fchollet) [inline keras/models.py]: Here: building `models.Sequential` with `Model` and `containers.Sequential`.

**Comment by fchollet** [comment]
> name is a string that internally maps to a layer number.

Support for naming layers is definitely something we'll add. But as much as possible we'd like it to stay an optional parameter.

> prev is the string that internally maps to the layer that current layer should be connected to, also internally mapped to a layer number. Defaults the last added layer.

What is the motivation for this, concretely? 

> Added LRN Layer
> Added Pooling with Strides as supported by Theano 0.7
> Added ZeroPadding2D Layer

We can definitely add these.

File: keras/layers/normalization.py [hunk 10]
```diff
@@ -64,6 +65,42 @@ def get_output(self, train):
...
+    """
+
+    def __init__(self, alpha=1e-4, k=2, beta=0.75, n=5, name=None, prev=None):
+        super(LRN2D, self).__init__(name, prev)
+        self.alpha = alpha
+        self.k = k
+        self.beta = beta
+        self.n = n
+        if n % 2 == 0:
+            raise NotImplementedError("Only works with odd n")
```
Comment (fchollet) [inline keras/layers/normalization.py]: This layer does not expose an input parameter; is it shape-agnostic?

**Comment by fchollet** [comment]
I think there seems to be good ideas in the PR, but it requires more work in terms of code quality. I am also not in line with these architectural changes.

A few questions:
- How exactly does the name-based connection scheme work?
- What is your motivation for this scheme; why do you think it was necessary over all over possible options?
- Have you considered adding the name-based connection scheme as a new class of model, or a new class of container? I don't think it makes sense to have it as part of Sequential (since it isn't sequential at all). I also think it is implementable completely independently of the layers. Layers have their own API that is independent from connection models. You can use this API to build any kind of new model / container. 

We won't merge the changes in their current state. You can fast-track the merge of LRN, Pooling with strides, and ZeroPadding by submitting them in a different PR, if you want.

**Comment by pranv** [comment]
The name based connection scheme is optional and really simple. 

In `Sequential`, layers are stacked up one over another. This is implemented in the form of an array `self.layers`. When a new layer is added, it's connected to the most recently added layer.This is done by the following line of code:
```
self.layers[-1].connect(self.layers[-2])
```

The previous layer's output is given as input. As done by these lines in core `Layer` class: 
```
def get_input(self, train):
        if hasattr(self, 'previous'):
            return self.previous.get_output(train=train)
        else:
            return self.input
```


Now, non-sequential models have 2 main differences from this Sequential model:
* Any newly added layer could be connected to any previously added layer[not just the most recently added layer]
* Layers take in multiple inputs. Multiple outputs(or fork) means that 2 or more layers obtain their input from the same layer. In other words, these layers are connected to the same layer.

In order to allow this flexibility along these two fronts, the user must be basically able to connect any layer to any layer in the model so far. Obviously, asking them to remember the layer numbers(basically the index of the layer in `self.layers`) is bad. No one can do that. Hence, I added a scheme where the layers have a string `name` that is mapped to a layer in `self.layers` by using `self.layer_nb` dictionary.
The previous layer, which will be giving it the input to the current layer can be accessed by this string.
Hence the `add` functions is modified as:
```
def add(self, layer):
        self.layers.append(layer)

        if layer.name is not None:
            self.layer_nb[layer.name] = len(self.layers) - 1

        if len(self.layers) > 1:
            if layer.prev_name is None:
                prev_layers = [self.layers[-2]] # no previous layer name provided, layer is connected to last added layer
            else:
                prev_layers = []
                for prev_layer in layer.prev_name:
                    prev_layer_nb = self.layer_nb[prev_layer]
                    prev_layers.append(self.layers[prev_layer_nb])

            self.layers[-1].connect(prev_layers)

```

And in the connect method of each layer:
```
def connect(self, node):
    self.previous = node #node is a list of previous layers.
    if len(node) == 1:
        self.previous = self.previous[0]    # if only one input (for backward compatibility)
```

Each layer has an optional `name` attribute and an optional `prev` attribute.

* `name` is a string that internally maps to a layer number.
* `prev` is the string that internally maps to the layer that current layer should be connected to, also internally mapped to a layer number. Defaults the last added layer.

The `Merge` layer gets it as a list of layers and takes input from all of them, while the rest of the layers that get a single node as input.

This will be made clear by this example. Consider the inception module used in iconic GoogLeNet:

inception


The Code to realize that module would be:
```
# The bottom block where the module starts
model.add(MaxPooling2D(poolsize=(2, 2), name='base'))


# 1st row after MaxPool
model.add(Convolution2D(64, 64, 1, 1, name='conv11', prev='base')
model.add(Convolution2D(64, 64, 1, 1, name='conv12', prev='base')
model.add(MaxPooling2D(poolsize=(3, 3), name='pool1, prev='base'))


# 2nd row after MaxPool
model.add(Convolution2D(64, 64, 1, 1, name='conv21', prev='base')

model.add(Convolution2D(64, 64, 3, 3, name='conv22', prev='conv11')
model.add(Convolution2D(64, 64, 5, 5, name='conv23', prev='conv12')
model.add(Convolution2D(64, 64, 1, 1, name='conv24', prev='pool1'))


# DEPTHCONCAT [Green Box]
model.add(Merge(prev['conv21', 'conv22', 'conv23', 'conv24' ], mode='concat'))

# Used one standard number of filters (64) for simplicity. The actual inception module has varying number of filters for each type of convolution.
```

I hope that the above example gives you a clear idea as to what role the 2 attributes play. Note that the `base` layer has no `prev` attribute. The final `Merge` layer has neither `name` nor `prev` attributes. They are entirely optional and the whole thing is backward compatible.

As far as code quality goes - yeah, it needs improvement and it was largely involuntary. I misunderstood `containers`, hence I changed that part. I'll revert that, fix the style violations at a couple of places and submit it. OR, if you think that this cannot be merged as a whole piece, I'll split it up

**Comment by iskandr** [comment]
As I'm getting into the Keras codebase (and this PR), I'm finding a confusing degree of polymorphism where an argument can either be a single object or a tuple of objects (e.g. `previous` is an object *or* a tuple). I think it's worth clarifying the expected types of all arguments in NumPy-style docstring parameter lists and explaining when/why an argument can be of multiple types.

**Comment by iskandr** [comment]
What's the testing situation in Keras? I don't see any unit tests in this PR.

**Comment by fchollet** [comment]
@pranv Thank you for the explanation. I do believe that such an arbitrary-connectivity scheme would be very valuable to have in Keras. And achieving it using unique identifiers for each layer sounds reasonable. 

However, the feature should not affect layers implementation, and it should be implemented as a container instead.

There are 3 levels of organization in Keras:
- layer: a self-contained computing unit that can be manipulated through `connect`, `get_ouput`, etc. 
- container: an input-to-output computation graph made of layers. 
- model: adds training/testing loop handlers on top of a container. 

New ways to connect layers together, like this one, should be introduced as a container.

In terms of API, you would have something like:
```python
# here model would be an instance of a class that inherits from Model and from your new container
model.add(Convolution2D(64, 64, 1, 1), name='conv11', input='base')
```

> OR, if you think that this cannot be merged as a whole piece, I'll split it up

You can submit the easy-to-merge changes separately, like LRN, zero padding, pooling with strides.

**Comment by fchollet** [comment]
> I'm finding a confusing degree of polymorphism where an argument can either be a single object or a tuple of objects (e.g. previous is an object or a tuple).

This should definitely be avoided. Where exactly have you found this outside of this PR?

**Comment by fchollet** [comment]
By the way, note that containers implement the layer API as well (they behave like meta-layers made out of layers).

Let's start a discussion about the implementation of a graph container (API and architecture). We can use your work on this PR as a basis to start on.

**Comment by iskandr** [comment]
@fchollet I think only in the arguments to `train` (https://github.com/fchollet/keras/blob/master/keras/models.py#L137), where `X` can be an array or a list of arrays. 

Also, I think once we have a Graph API then we'll have an even richer set of possible inputs there (e.g. multiple named input and output arrays), not sure how to keep the possibilities clean and understandable.

**Comment by fchollet** [comment]
@iskandr user-facing input *should* often be polymorphic. That's good UX. Theano does this a lot as well. Resolving the type is a matter of input validation (which should *always* happen on user-provided inputs anyway).

Anything that's not user-facing input should definitely not be polymorphic, though.

**Comment by iskandr** [comment]
@fchollet I agree and I have no trouble with the inputs to `train` being polymorphic. However, as a new reader of this codebase I found the lack of documentation a little confusing. Maybe a docstring would help here? 

I assumed when I saw more polymorphism in this PR and the other PR for input/output dimension computation that maybe it was just a repo-wide preference to allow implicit sequences, happy that's not the case!

**Comment by pranv** [comment]
@iskandr, both the places where I have used Polymorphism are user facing and they are the right ways to do it. It is the spirit and essence of Python and that's why people use it. More importantly, it makes things easier to use and understand too. Maybe a little documentation was needed, but not really a major roadblock.

**Comment by jramapuram** [comment]
Any thoughts about extending Graphs to multi output type solutions? Eg: Use an encoder/decoder framework coupled with Attention as an example. Lots of cool stuff at ICML this year with these model paradigms.

**Comment by fchollet** [comment]
> Any thoughts about extending Graphs to multi output type solutions?

The Graph model that was released last week does support multiple outputs (with a specific loss function for each output), if that's what you mean.

**Comment by jramapuram** [comment]
Awesome! Thank you!
